src/GraphBLAS-sharp.Backend/Algorithms/PageRank.fs

```diff
@@ -0,0 +1,198 @@
+
+            match matrix with
+            | ClMatrix.CSR matrix ->
+
+                let outDegree = countOutDegree queue matrix
+
+                let resultValues =
+                    clContext.CreateClArrayWithSpecificAllocationMode(DeviceOnly, matrix.Values.Length)
+
+                let kernel = multiply.GetKernel()
+
+                let ndRange =
+                    Range1D.CreateValid(matrix.RowCount * workGroupSize, workGroupSize)
+
+                queue.Post(
+                    Msg.MsgSetArguments
+                        (fun () ->
+                            kernel.KernelFunc
+                                ndRange
+                                matrix.RowCount
+                                matrix.RowPointers
+                                matrix.Values
+                                outDegree
+                                resultValues)
+                )
+
+                queue.Post(Msg.CreateRunMsg<_, _> kernel)
+
+                outDegree.Free queue
+
+                let newMatrix =
+                    { Context = clContext
+                      RowCount = matrix.RowCount
+                      ColumnCount = matrix.ColumnCount
+                      RowPointers = copy queue DeviceOnly matrix.RowPointers
+                      Columns = copy queue DeviceOnly matrix.Columns
+                      Values = resultValues }
+
+                transposeInPlace queue DeviceOnly newMatrix
+                |> ClMatrix.CSR
+                |> PageRankMatrix
+            | _ -> failwith "Not implemented"
+
+    let run (clContext: ClContext) workGroupSize =
+
+        let squareOfDifference = ArithmeticOperations.squareOfDifference
+        let plus = ArithmeticOperations.float32SumOption
+        let mul = ArithmeticOperations.float32MulOption
+
+        let spMVTo =
+            Operations.SpMVInPlace plus mul clContext workGroupSize
+
+        let addToResult =
+            GraphBLAS.FSharp.Vector.map2InPlace plus clContext workGroupSize
+
+        let subtractAndSquare =
+            GraphBLAS.FSharp.Vector.map2To squareOfDifference clContext workGroupSize
+
+        let reduce =
+            GraphBLAS.FSharp.Vector.reduce <@ (+) @> clContext workGroupSize
+
+        let create =
+            GraphBLAS.FSharp.Vector.create clContext workGroupSize
+
+        fun (queue: MailboxProcessor<Msg>) (PageRankMatrix matrix) accuracy ->
+            let vertexCount = matrix.RowCount
+
+            //None is 0
+            let mutable rank =
+                create queue DeviceOnly vertexCount Dense None
+
+            let mutable prevRank =
+                create queue DeviceOnly vertexCount Dense (Some(1.0f / (float32 vertexCount)))
+
+            let mutable errors =
+                create queue DeviceOnly vertexCount Dense None
+
+            let addition =
+                create queue DeviceOnly vertexCount Dense (Some((1.0f - alpha) / (float32 vertexCount)))
+
+            let mutable error = accuracy + 0.1f
```

Comment:
I'm all about a separate module for constants. Then it would be possible to take out actions like reciprocal and the like. Then, perhaps such actions will look more meaningful.

Reply:
Or use predefined ones. Or comment on it. It's just that the number of magic constants scares me.

Reply:
It still confuses me. And float arithmetic 3 line above too!